core/services/synchronization/telemetry_ingress_client.go

```diff
@@ -0,0 +1,228 @@
+}
+
+// Close disconnects the wsrpc client from the ingress server
+func (tc *telemetryIngressClient) Close() error {
+	return tc.StopOnce("TelemetryIngressClient", func() error {
+		close(tc.chDone)
+		tc.wgDone.Wait()
+		return nil
+	})
+}
+
+func (tc *telemetryIngressClient) connect(clientPrivKey []byte) {
+	tc.wgDone.Add(1)
+
+	go func() {
+		defer tc.wgDone.Done()
+
+		serverPubKey := keys.FromHex(tc.serverPubKeyHex)
+
+		conn, err := tc.wsrpcDial(tc.url.String(), wsrpc.WithTransportCreds(clientPrivKey, serverPubKey))
+		if err != nil {
+			logger.Errorf("Error connecting to telemetry ingress server: %v", err)
+			return
+		}
+		defer conn.Close()
+
+		// Initialize a new wsrpc client caller
+		// This is used to call RPC methods on the server
+		tc.mu.Lock()
+		tc.wsrpcClient = tc.newTelemClient(conn)
+
+		// Start handler for telemetry
+		tc.handleTelemetry()
+
+		tc.isReady = true
+		tc.mu.Unlock()
+
+		// Wait for close
+		<-tc.chDone
+
+	}()
+}
+
+func (tc *telemetryIngressClient) handleTelemetry() {
+	go func() {
+		for {
+			select {
+			case p := <-tc.chTelemetry:
+				// Send telemetry to the ingress server, log any errors
+				telemReq := &telemPb.TelemRequest{Telemetry: p.Telemetry, Address: p.ContractAddress.String()}
+				_, err := tc.wsrpcClient.Telem(p.Ctx, telemReq)
+				if err != nil {
+					logger.Errorf("Could not send telemetry: %v", err)
+					continue
+				}
+				if tc.logging {
+					logger.Debugw("successfully sent telemetry to ingress server", "contractAddress", p.ContractAddress.String(), "telemetry", p.Telemetry)
+				}
+			case <-tc.chDone:
+				return
+			}
+		}
+	}()
+}
+
+// logBufferFullWithExpBackoff logs messages at
+// 1
+// 2
+// 4
+// 8
+// 16
+// 32
+// 64
+// 100
+// 200
+// 300
+// etc...
+func (tc *telemetryIngressClient) logBufferFullWithExpBackoff(payload TelemPayload) {
+	count := atomic.AddUint32(&tc.dropMessageCount, 1)
+	if count > 0 && (count%100 == 0 || count&(count-1) == 0) {
+		logger.Warnw("telemetry ingress client buffer full, dropping message", "telemetry", payload.Telemetry, "droppedCount", count)
+	}
+}
+
+// getCSAPrivateKey gets the client's CSA private key
+func (tc *telemetryIngressClient) getCSAPrivateKey() (privkey []byte, err error) {
+	// Fetch the client's public key
+	keys, err := tc.ks.ListCSAKeys()
+	if err != nil {
+		return privkey, err
+	}
+	if len(keys) < 1 {
+		return privkey, errors.New("CSA key does not exist")
+	}
+
+	privkey, err = tc.ks.Unsafe_GetUnlockedPrivateKey(keys[0].PublicKey)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	return privkey, nil
+}
+
+// Send sends telemetry to the ingress server using wsrpc if the client is ready.
+// Also stores telemetry in a small buffer in case of backpressure from wsrpc,
+// throwing away messages once buffer is full
+func (tc *telemetryIngressClient) Send(payload TelemPayload) {
+	tc.mu.RLock()
+	defer tc.mu.RUnlock()
+
+	if !tc.isReady {
+		logger.Error("Could not send telemetry, client is not ready")
+		return
+	}
```

Comment:
@j16r Do we need a mutex here? The `isReady` bool only gets written once by a single go routine and is read by all others. I used a mutex to be cautious, but I'm wondering if there's a subtle exception here

Reply:
You shouldn't need it.

The writer will lock it so it will wait on `tc.isReady` read until the lock is released. All this lock will do is prevent concurrent `Send` calls from executing at the same time as they have to wait for the first Send to release the read lock before it can continue.